src/coreclr/vm/methodtablebuilder.cpp

```diff
@@ -7298,16 +7298,14 @@ VOID MethodTableBuilder::AllocAndInitMethodDescChunk(COUNT_T startIndex, COUNT_T
             // and should not be used.  We should go to the effort of having proper constructors
             // in the MethodDesc class. </NICE>
 
-            memcpy(pUnboxedMD, pMD, pMD->SizeOf());
-
-            // Reset the chunk index
-            pUnboxedMD->SetChunkIndex(pChunk);
-
-            if (bmtGenerics->GetNumGenericArgs() == 0) {
-                pUnboxedMD->SetHasNonVtableSlot();
+            if (bmtGenerics->GetNumGenericArgs() == 0)
+            {
+                memcpy(pUnboxedMD, pMD, pMD->GetBaseSize());
```

Comment:
This is basically the fix. `memcpy(pUnboxedMD, pMD, pMD->SizeOf());` vs. `memcpy(pUnboxedMD, pMD, pMD->GetBaseSize());`. 

`memcpy(pUnboxedMD, pMD, pMD->SizeOf());` may copy a few fields to wrong locations. Then it is not enough to write correct values to the new locations, we would also need to clean up wrong locations.

It is better for the initial memcopy to do only the base portion, which is the same between different layouts. The rest stays zero-inited.

Reply:
The problem that we see now is that we are corrupting a slot that is later set via interlocked exchange, which expects 0 comparand. 
Since we have junk in the slot, the exchange does not happen and the junk stays in the slot and causes crashes.